### DISCUSSION TITLE
To modify default HTML template on-the-fly

### DISCUSSION THREAD

**Discussion by johnmapeson** [jgm/pandoc#9788] [Q&A] [answered]
Is it possible to modify the default HTML template on-the-fly, without using a file? Specifically, I need to strip out this line:

```
<title>$if(title-prefix)$$title-prefix$ – $endif$$pagetitle$</title>
```

Is it possible to do using something like the example below?

```
pandoc input.md -s --template=default | sed ... -o output.htm 
```

**Comment by johnmapeson** [comment]
Okay, to answer my own question. To store contents of the default template in a variable and strip the `<title>` line from there, one can use

```
template=$(grep -v '<title>' <<< $(pandoc -D html))
```

or, if using Zsh,

```
template=${(F)"${(f@)$(pandoc -D html)}":#*<title>*}
```

But then we need to pass the resulting string as a `--template` value. I will open a separate discussion about this.

**Comment by tarleb** [comment, marked as answer]
Alternative approach: use a custom Lua writer and remove the title line in Lua.

``` lua
Extensions = pandoc.format.extensions 'html5'
Template = pandoc.template.default('html5'):gsub('\n<title>[^\n]*', '')
Writer = function (doc, opts)
  return pandoc.write(doc, {format='html5', extensions=opts.extensions}, opts)
end
```

**Reply by johnmapeson** [reply]
Many thanks, Albert!

**Reply by johnmapeson** [reply]
Because the `<title>` line is indented, the regex should also match spaces before it. I tried `%s` (and, of course, `\s*`), which should match zero or more spaces, but, for some reason, it doesn't work for me. Could you point what is wrong here? I never worked with Lua-flavored regular expressions before.

```
gsub('\n%s<title>[^\n]*', '')
```

Edit: Nevermind, it works.

**Reply by johnmapeson** [reply]
Albert, maybe you know, is it possible to use this code not as a custom writer, but as a regular Lua filter?

**Reply by tarleb** [reply]
I don't think it's possible¹ to do that in a filter, as only the writer can modify the template.

¹ or rather *wise*, in that a lot of complicated hacks could be thought up, but none that seem sensible.

**Comment by johnmapeson** [comment]
@tarleb Hello, Albert. I know this is an old thread, and this is not related to Pandoc directly, but could you explain why you have used `Writer = function (doc, opts)` instead of `Writer (doc, opts)`?

Another example, from <https://github.com/jgm/pandoc/discussions/9697>. Your code:

```
local hex_to_char = function(x)
  return string.char(tonumber(x, 16))
end

local unescape = function(url)
  return url:gsub("%%(%x%x)", hex_to_char)
end

function Image (img)
  if not (img.src:match('^https://') or img.src:match('%.%.')) then
    return nil
  end

  local mt, contents = pandoc.mediabag.fetch(img.src)
  img.src = pandoc.path.filename(unescape(img.src))
  pandoc.mediabag.insert(img.src, mt, contents)

  return img
end
```

Why you have used `local hex_to_char = function(x)` and `local unescape = function(url)` instead of `local function hex_to_char(x)` and `local function unescape(url)`?

One another example, from <https://github.com/jgm/pandoc/discussions/9514>. Here you use `local function starts_with(start, str)`, and not `local starts_with = function(start, str)`:

```
local function starts_with(start, str)
  return str:sub(1, #start) == start
end

function RawInline (el)
  return starts_with('<!--', el.text)
    and pandoc.RawInline('markdown', el.text)
    or {}
end
```

Asking such silly questions is my regular way to grok into something new to me, like a new programming language. :)

**Reply by tarleb** [reply]
The `Writer = ...` is just for symmetry with the other definitions. It just looks slightly more pleasant to my eyes. There's no deeper meaning.

It's similar for the other code snippets: I sometimes use the alternative syntax to mark some functions as being different than others. E.g., to make the distinction of helper functions and filter functions more obvious. But it's not a consistent convention, as you noticed.
